### ISSUE TITLE
(Unintended) Destruction of Thread object causes hard to debug crash

### ISSUE THREAD

**Issue by skunkiferous** [nim-lang/Nim#7172]
Me, and people trying to help me on the forum, spent a while to debug a crash, which was due to a Thread object being destroyed, while the thread is still running. In this particular case, I believe the Thread object was on the stack, but I assume the same would happen with a GCed Thread object.

Here is a simple example that causes the crash, at least for me on Windows, and was also reproducible on Archlinux:

```nim
import os

proc whatever() {.thread, nimcall.} =
  echo("TEST")

proc initProcess(): void =
  echo("In initProcess()")
  var thread: Thread[void]
  createThread(thread, whatever)
  echo("initProcess() done")

proc doStuff(): void =
  echo("In doStuff()")
  # ...
  initProcess()
  sleep(500)
  # ...
  echo("Crashes before getting here!")

doStuff()
```

If "var thread" is moved to the global scope, the crash does not happen.

It is not clear to me if the crash can be prevented though a compile-time error, but at least it would be good to have an error message that users understand. It might also be good to mention that this can happen in the Thread and/or createThread documentation.

**Comment by cheatfate** [comment]
@andreaferretti, please unmark `Threads`, because this is not bug of threads. If people creates something in stack, they need to be ready that this will be lost after procedure exit.

**Comment by andreaferretti** [comment]
I agree it is not a bug of threads - in some sense it is not a bug at all. Still I thought it would be useful  to locate issues such as this when looking for thread-related problems

**Comment by konqoro** [comment]
Doesnt crash for me. Nim latest head.

**Comment by Yardanico** [comment]
Crashes for me on Arch Linux with GCC 7.3
@konqoro what is your platform/compiler?

**Comment by cheatfate** [comment]
@konqoro, please run many times executable, crashes can be not constant...

**Comment by konqoro** [comment]
Tried again and without '-d:release' crashes every single time. with release doesn't. gcc version 7.2.1 20180116 (GCC) manjaro

**Comment by genotrance** [comment]
Confirmed works on Windows, crashes on Ubuntu.

**Comment by narimiran** [comment]
>  without '-d:release' crashes every single time

I've run it several times (without `-d:release`) on Manjaro Linux, with Nim 1.5, and it never crashed.

**Comment by kobi2187** [comment]
I think that it is not clear from the docs what happens on createThread.
If the thread instance is added to some global sequence, and can be retrieved later, then I guess it shouldn't crash. But if you're trying to run an operation through an object which by this time is already nil, you will get a runtime error, in other words, a crash. Should the compiler warn? usually it doesn't know about runtime state.
